File: TheaterSuggestions/Java/ScreeningSeatingArrangementSuggestions/src/main/java/org/octo/seatingplacessuggestions/seatingplacesuggestions/Row.java
package org.octo.seatingplacessuggestions.seatingplacesuggestions;

import java.util.ArrayList;
import java.util.List;

public record Row(String name, List<SeatingPlace> seatingPlaces) {

    public SeatingOptionIsSuggested suggestSeatingOption(int partyRequested, PricingCategory pricingCategory) {

        var seatAllocation = new SeatingOptionIsSuggested(partyRequested, pricingCategory);

        for (var seat : seatingPlaces) {
            if (seat.isAvailable() && seat.matchCategory(pricingCategory)) {
                seatAllocation.addSeat(seat);

                if (seatAllocation.matchExpectation())
                    return seatAllocation;

            }
        }
        return new SeatingOptionIsNotAvailable(partyRequested, pricingCategory);
    }

    public Row addSeatingPlace(SeatingPlace seatingPlace) {
        List<SeatingPlace> newSeatingPlaces = new ArrayList<>( seatingPlaces );
        newSeatingPlaces.add(seatingPlace);
        return new Row(name, newSeatingPlaces);
    }

    public Row allocate(SeatingPlace seatingPlaceSuggested) {
        ArrayList<SeatingPlace> newSeatingPlaces = new ArrayList<>();
        seatingPlaces.forEach(seatingPlace -> {
            if (seatingPlaceSuggested.isSameLocation(seatingPlace))
                newSeatingPlaces.add(seatingPlaceSuggested.allocate());
            else
                newSeatingPlaces.add(seatingPlace);
        });
        return new Row(name, newSeatingPlaces);
    }

}

File: TheaterSuggestions/CSharp/ScreeningSeatingArrangementSuggestions/Row.cs
using Value;

namespace SeatsSuggestions;

public class Row(string name, List<SeatingPlace> seats) : ValueType<Row>
{
    public string Name { get; } = name;
    public List<SeatingPlace> SeatingPlaces { get; } = seats;

    public SeatingOptionIsSuggested SuggestSeatingOption(int partyRequested, PricingCategory pricingCategory)
    { 
        var seatAllocation = new SeatingOptionIsSuggested(partyRequested, pricingCategory);
        
        foreach (var seat in SeatingPlaces)
        {
            if (seat.IsAvailable() && seat.MatchCategory(pricingCategory))
            {
                seatAllocation.AddSeat(seat);
                if (seatAllocation.MatchExpectation()) return seatAllocation;
            }
        }
        return new SeatingOptionIsNotAvailable(partyRequested, pricingCategory);
    }

    public Row AddSeatingPlace(SeatingPlace seatingPlace)
    {
        return new Row(Name, [..SeatingPlaces, seatingPlace]);
    }

    protected override IEnumerable<object> GetAllAttributesToBeUsedForEquality()
    {
        return [Name, new ListByValue<SeatingPlace>(SeatingPlaces)];
    }

    public Row Allocate(SeatingPlace seatingPlacesSuggested)
    {
        var seatingPlaces = new List<SeatingPlace>();
        foreach (var seatingPlace in SeatingPlaces)
        {
            if (seatingPlacesSuggested.IsSameLocation(seatingPlace))
            {
                seatingPlaces.Add(seatingPlacesSuggested.Allocate());
            }
            else
            {
                seatingPlaces.Add(seatingPlace);
            }
        }
        return new Row(Name, seatingPlaces);
    }
}